### PR TITLE
Implement navigator.deviceMemory farbling

### PR DESCRIPTION
**Pull request by pilgrim-brave**
Resolves https://github.com/brave/brave-browser/issues/12348

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [ ] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [ ] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Review by fmarier — APPROVED** [review]
I didn't look closely through the whole patch, but the approach LGTM.

File: test/BUILD.gn [hunk 1]
```diff
@@ -514,6 +514,7 @@ if (!is_android) {
       "//brave/browser/extensions/brave_extension_provider_browsertest.cc",
       "//brave/browser/extensions/brave_theme_event_router_browsertest.cc",
       "//brave/browser/farbling/brave_enumeratedevices_farbling_browsertest.cc",
+      "//brave/browser/farbling/brave_navigator_devicememory_farbling_browsertest.cc",
```
Comment (mkarolin) [inline test/BUILD.gn:517]: We no longer add tests directly to test/BUILD.gn. New tests need to go to their own source_set in the directory where they live and then added as a dep here.

File: browser/farbling/brave_navigator_devicememory_farbling_browsertest.cc [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  int ExecScriptGetInt(const std::string& script, T* frame) {
+    int value;
+    EXPECT_TRUE(ExecuteScriptAndExtractInt(frame, script, &value));
+    return value;
+  }
+
+  template <typename T>
+  std::string ExecScriptGetStr(const std::string& script, T* frame) {
+    std::string value;
+    EXPECT_TRUE(ExecuteScriptAndExtractString(frame, script, &value));
```
[the remark this answers is not in the document]
Comment (pilgrim-brave) [reply browser/farbling/brave_navigator_devicememory_farbling_browsertest.cc]: latest push uses `EvalJs`

File: test/BUILD.gn (hunk 1 above)
Comment (pilgrim-brave) [reply test/BUILD.gn]: latest push has a `source_set("browser_tests")` in `browser/farbling/BUILD.gn`

**Review by mkarolin — APPROVED** [review]
LGTM

**Pull request merged**